pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -355,55 +358,97 @@ public void removeProducer(Producer producer) {
             lock.readLock().unlock();
         }
 
+        CompletableFuture<Subscription> subscriptionFuture = isDurable ? //
+                getDurableSubscription(subscriptionName) //
+                : getNonDurableSubscription(subscriptionName, startMessageId);
+
+        int maxUnackedMessages  = isDurable ? brokerService.pulsar().getConfiguration().getMaxUnackedMessagesPerConsumer() :0;
+
+        subscriptionFuture.thenAccept(subscription -> {
+            try {
+                Consumer consumer = new Consumer(subscription, subType, consumerId, priorityLevel, consumerName,
+                        maxUnackedMessages, cnx, cnx.getRole());
+                subscription.addConsumer(consumer);
+                if (!cnx.isActive()) {
```

Comment:
This part of the code is exactly the same as before, I just refactored the way to get the `subscriptionFuture`. For the specific concern, the problem is that the connection can flip to closed even after we have checked the `isActive()`, so we need to double check anyway the connection status *after* we have created the consumer, which internally register itself to get notified when the connection is gone.